tests/functional/lib/BackbeatConsumer.js

```diff
@@ -0,0 +1,221 @@
+    const groupIdConc = `replication-group-deferred-${Math.random()}`;
+    let producer;
+    let consumer;
+    let consumedMessages = [];
+
+    function queueProcessor(message, cb) {
+        consumedMessages.push(message.value);
+        if (JSON.parse(message.value.toString()).deferred) {
+            process.nextTick(() => cb(null, { committable: false }));
+            setTimeout(() => {
+                consumer.onEntryCommittable(message);
+            }, 900 + Math.floor(Math.random() * 200));
+        } else {
+            process.nextTick(cb);
+        }
+    }
+    before(function before(done) {
+        this.timeout(60000);
+
+        producer = new BackbeatProducer({
+            kafka: kafkaConf,
+            topic: topicConc,
+            pollIntervalMs: 100,
+        });
+        consumer = new BackbeatConsumer({
+            zookeeper: zookeeperConf,
+            kafka: kafkaConf, groupId: groupIdConc, topic: topicConc,
+            queueProcessor,
+            concurrency: 10,
+            bootstrap: true,
+        });
+        async.parallel([
+            innerDone => producer.on('ready', innerDone),
+            innerDone => consumer.on('ready', innerDone),
+        ], done);
+    });
+    afterEach(() => {
+        consumedMessages = [];
+        consumer.removeAllListeners('consumed');
+    });
+    after(function after(done) {
+        this.timeout(10000);
+        async.parallel([
+            innerDone => producer.close(innerDone),
+            innerDone => consumer.close(innerDone),
+        ], done);
+    });
+
+    it('should be able to process 1000 messages with some deferred ' +
+    'committable status', done => {
+        const boatloadOfMessages = [];
+        for (let i = 0; i < 1000; ++i) {
+            boatloadOfMessages.push({
+                key: `message-${i}`,
+                message: `{"message_index":"${i}",` +
+                    `"deferred":${i % 2 === 0 ? 'true' : 'false'}}`,
+            });
+        }
+        setTimeout(() => producer.send(boatloadOfMessages, err => {
+            assert.ifError(err);
+        }), 1000);
+        let totalConsumed = 0;
+        consumer.subscribe();
+        consumer.on('consumed', messagesConsumed => {
+            totalConsumed += messagesConsumed;
+            assert(totalConsumed <= boatloadOfMessages.length);
+            if (totalConsumed === boatloadOfMessages.length) {
+                assert.strictEqual(
+                    consumer.getOffsetLedger().getProcessingCount(),
+                    500);
+                // offsets are set to be committable after 1 second in this
+                // test, so wait for 2 seconds
+                setTimeout(() => {
+                    assert.strictEqual(
+                        consumer.getOffsetLedger().getProcessingCount(),
+                        0);
+                    done();
+                }, 2000);
+            }
+        });
+    }).timeout(30000);
```

Comment:
For timeouts, let's set them globally at once. It can be done in package.json when invoked by command or mocha.opts.